src/supply_info.rs

```diff
@@ -0,0 +1,99 @@
+//! Structs and logic related to supply information management for assets.
+
+use std::collections::{hash_map, HashMap};
+
+use crate::{issuance::Error, note::AssetBase, value::ValueSum};
+
+/// Represents the amount of an asset and its finalization status.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct AssetSupply {
+    /// The amount of the asset.
+    pub amount: ValueSum,
+    /// Whether or not the asset is finalized.
+    pub is_finalized: bool,
+}
+
+impl AssetSupply {
+    /// Creates a new AssetSupply instance with the given amount and finalization status.
+    pub fn new(amount: ValueSum, is_finalized: bool) -> Self {
+        Self {
+            amount,
+            is_finalized,
+        }
+    }
+}
+
+/// Contains information about the supply of assets.
+#[derive(Debug, Clone, Default)]
+pub struct SupplyInfo {
+    /// A map of asset bases to their respective supply information.
+    pub assets: HashMap<AssetBase, AssetSupply>,
+}
+
+impl SupplyInfo {
+    /// Inserts or updates an asset's supply information in the supply info map.
+    /// If the asset exists, adds the amounts (unconditionally) and updates the finalization status
+    /// (only if the new supply is finalized). If the asset is not found, inserts the new supply.
+    pub fn add_supply(&mut self, asset: AssetBase, new_supply: AssetSupply) -> Result<(), Error> {
+        match self.assets.entry(asset) {
+            hash_map::Entry::Occupied(mut entry) => {
+                let supply = entry.get_mut();
```

Comment:
Somthing strange here, why we need `.get_mut()` if `entry` is already `mut`?

Reply:
In general, I am using the `entry()` method now to avoid calling both `get` and `insert`. There is a more elegant way to work with `entry` result, using `and_update` and `or_insert`, but `and_update` does not allow returning an error. So, for our current logic, using a `match` with `entry` is the optimal choice. `get_mut` does require a mutable `self` reference, so `entry` needs to be mutable. However, in this particular case, we only need to use `entry` once, so we can convert it into `AssetSupply` using the `into_mut` method:

`
hash_map::Entry::Occupied(entry) => {
    let supply = entry.into_mut();
    supply.amount =
        (supply.amount + new_supply.amount).ok_or(Error::ValueSumOverflow)?;
    supply.is_finalized |= new_supply.is_finalized;
}
`

I have fixed it now.